File: imsdia/api/c++/mtest/diameter/s6a/src/CancelLocationTest.cpp
/********************************-*-CPP-*-***********************************
 *                                                                          *
 *     Copyright 2006 IntelliNet Technologies, Inc. All Rights Reserved.    *
 *             Manufactured in the United States of America.                *
 *       1990 W. New Haven Ste. 312, Melbourne, Florida, 32904 U.S.A.       *
 *                                                                          *
 *   This product and related documentation is protected by copyright and   *
 *   distributed under licenses restricting its use, copying, distribution  *
 *   and decompilation.  No part of this product or related documentation   *
 *   may be reproduced in any form by any means without prior written       *
 *   authorization of IntelliNet Technologies and its licensors, if any.    *
 *                                                                          *
 *   RESTRICTED RIGHTS LEGEND:  Use, duplication, or disclosure by the      *
 *   government is subject to restrictions as set forth in subparagraph     *
 *   (c)(1)(ii) of the Rights in Technical Data and Computer Software       *
 *   clause at DFARS 252.227-7013 and FAR 52.227-19.                        *
 *                                                                          *
 ****************************************************************************
 *                                                                          *
 * CONTRACT: INTERNAL                                                       *
 ****************************************************************************
 * LOG: $Log: CancelLocationTest.cpp,v $
 * LOG: Revision 1.3  2009/03/04 12:43:05  sureshj
 * LOG: Demo changes for Motorola specific part changes.TS 29272 v8.1.0.
 * LOG:
 * LOG: Revision 1.2  2008/11/12 05:42:42  sureshj
 * LOG: Updates for supporting v8.0.0. part of release 3.1.1.
 * LOG:
 * LOG: Revision 1.1  2008/08/19 06:58:33  sureshj
 * LOG: Baseline demo application for s6a interface.
 * LOG:							    	
 * ID: $Id: CancelLocationTest.cpp,v 1.3 2009/03/04 12:43:05 sureshj Exp $								    	
 ****************************************************************************/

#include <app.h>
#include <s6a_defines.h>
#include <s6a_stats.h>
#include <diameter/s6a/CancelLocation.h>

using namespace std;
using namespace its;
using namespace diameter;
using namespace diameter::base;
using namespace diameter::s6a;

extern string sessIdString;
extern ITS_UINT sIdx;

/*.implementation:public
 ****************************************************************************
 *  Purpose: This function sends the  CancelLocation to the
 *      diameter stack. Create the  CancelLocation object,
 *      populate the AVPs and send the message to the stack.
 *
 *  Input Parameters:
 *      Update choice.
 *
 *  Input/Output Parameters:
 *      None.
 *
 *  Output Parameters:
 *      None.
 *
 *  Return Value:
 *
 *  Notes:
 *
 *  See Also:
 ****************************************************************************/
int SendS6ACancelLocationRequest()
{
    // [1]. Create an instance of CancelLocationRequest.
    // Optional param error - whether this command carries an error indicator.
    // Optional param retrans - whether this is a retransmission.
    diameter::s6a::CancelLocationRequest CLReq;

    // [2.1.1.x] Fixed AVP
    // [2.1.1.1] Set the Session-Id AVP.
    // The Session-Id AVP is used to identity a specific session
    // and must appear immediately following the Diameter header.
    // The AVP Code of Session-Id is 263
    // The data type of Session-Id is UTF8String.
    AllocateSession(S6A_INTERFACE_APP_ID, sIdx, sessIdString);
    CLReq.setSessionId(diameter::base::SessionId(sessIdString));

    CLReq.setSessionIndex(sIdx);

    // [2.1.2.x] Required AVPs
    // [2.1.2.2]Create a Auth-Session-State AVP object.
    // The Auth-Session-State AVP is specifies whether state is
    // maintained for a particular session.
    // The AVP Code of Auth-Session-State AVP is 277
    // The data type of Auth-Session-State AVP is Enumerated with
    // the following values
    // {
    //  STATE_MAINTAINED(0),
    //  NO_STATE_MAINTAINED(1)
    // }
    AuthSessionState authSessionState(AuthSessionState::NO_STATE_MAINTAINED);
    CLReq.setAuthSessionState(authSessionState);

    // [2.1.2.3] Set the Origin-Host AVP.
    // The Origin-Host AVP identifies the end-point that originated
    // the Diameter message.
    // The Origin-Host AVP must be present in all the Diameter messages.
    // This AVP SHOULD be placed as close to the Diameter header as possible.
    // The relay agents must not modify this AVP.
    // The AVP Code of Origin-Host AVP is 264
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginHost originHost(GetOriginHost());
    CLReq.setOriginHost(originHost);

    // [2.1.2.4] Set the Origin-Realm AVP.
    // The Origin-Realm AVP contains the realm of the originator of
    // any Diameter message and must be present in all the Diameter messages.
    // The AVP Code of Origin-Host AVP is 296.
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginRealm originRealm(GetOriginRealm());
    CLReq.setOriginRealm(originRealm);


    // [2.1.2.5] Set the Destination-Host AVP.
    // [2.2.2] Populate optional paramete "Destination-Host".
    // Set the Destination-Host AVP.
    // The data type of Destination-Host is DiameterIdentity
    // This AVP identifies the endpoint where  the Diameter message is destinated
    // This AVP SHOULD be placed as close to the Diameter header
    // The AVP Code of Destination-Host is 293.
    DestinationHost DestinationHost(GetDestinationHost());
    CLReq.setDestinationHost(DestinationHost);

    // [2.1.2.6] Set the Desti
    // The Destination-Realm AVP contains the realm, the message has
    // to be routed to.
    // The Destination-Realm AVP is used to perform message routing decisions.
    // The AVP Code of Destination-Realm AVP is 283
    // The data type of Destination-Realm AVP is DiameterIdentity.
    DestinationRealm destinationRealm(GetDestinationRealm());
    CLReq.setDestinationRealm(destinationRealm);

    // [2.1.2.7] Set the User-Name AVP.
    // Set the User-Name AVP.
    // The data type of User-Name is UTF8String.
    // This AVP contains the User-Name.
    // The AVP Code of User-Name is 1.
    UserName userName("sip:user@3gppnetwork.com");
    CLReq.setUserName(userName);

    // Set Cancelation-Type AVP .
     CancellationType cantype(CancellationType::SGSN_UPDATE_PROCEDURE);
     CLReq.setCancellationType(cantype);

    // Application should not set this AVP.
    // CLReq.addProxyInfo(diameter::base::ProxyInfo(...));
    // CLReq.addRouteRecord(diameter::base::RouteRecord(...));		   
    // add more AVP as ANY values

    if (appMode == UNIT_TEST)
    cout<<CLReq;

    // Send object to stack.
    SendDiaAppMsg(CLReq, sIdx);

    S6AStats::GetS6AStats()->UpdateSendStats(CLReq.getCommandCode(), ITS_TRUE);
    return ITS_SUCCESS;
}


/*.implementation:public
 ****************************************************************************
 *  Purpose:
 *      This function sends the Answer to the CancelLocation request
 *      made by the diameter stack. Create the CancelLocationAnswer
 *      object, populate the AVPs and send the message to the stack.
 *
 *  Input Parameters:
 *      const diameter::CommandImpl *impl: Pointer to the received Request
 *      Command, for which the Answer is to be sent.
 *
 *  Input/Output Parameters:
 *      None.
 *
 *  Output Parameters:
 *      None.
 *
 *  Return Value:
 *
 *  Notes:
 *
 *  See Also:
 ****************************************************************************/
int SendS6ACancelLocationAnswer(const CommandImpl * impl)
{
    // 1. Create an instance of CancelLocationAnswer class
    // The Update-Data-Answer command is sent by a server
    // in response to the Update-Data-Request command with
    // Command-Code = 306 and 'R' bit reset
    // Optional param error - whether this command carries an error indicator.
    // Optional param retrans - whether this is a retransmission.
    diameter::s6a::CancelLocationAnswer CLAns;

    // Create an instance of CancelLocationRequest class to get host and
    // session details
    CancelLocationRequest CLreq(*impl);

    // Get AVP values from requested message
    const SessionId &sid1 = CLreq.getSessionId();
    const OriginHost &oh1 = CLreq.getOriginHost();
    const OriginRealm &or1 = CLreq.getOriginRealm();
    const AuthSessionState &authSessState = CLreq.getAuthSessionState();

    // 2. Create the instances of the classes describing the various
    // AVPs' present in an MultimediaAuthAnswer command and
    // pass them by value to the corresponding data members set
    // methods of the MultimediaAuthAnswer object created
    // in step 1.
    // [2.1.1] Set the Session-Id AVP.
    // The Session-Id AVP is used to identity a specific session
    // and must appear immediately following the Diameter header.
    // The AVP Code of Session-Id is 263
    // The data type of Session-Id is UTF8String.
    CLAns.setSessionId(sid1);
    CLAns.setSessionIndex(CLreq.getSessionIndex());    
 
    // [2.1.2.x] Required AVPs
    CLAns.setAuthSessionState(authSessState);
     
    // [2.2] Set the Result-Code AVP .
    // The Result-Code AVP indicates whether a particular request
    // was completed successfully or whether an error occurred
    // All Diameter answer messages must contain one Result-Code AVP
    // Diameter provides the following classes of error which
    // is represented by the Result-Code AVP:
    // 1xxx - Informational
    // 2xxx - Success
    // 3xxx - Protocol Errors
    // 4xxx - Transient Failures
    // 5xxx - Permanent Failure
    // The AVP Code of Result-Code AVP is 268.
    // The data type of Result-Code AVP is Unsigned32.
    // already ExperimentalResult is been set
    ResultCode resultCode(2001u);
    CLAns.setResultCode(resultCode);

    // Create the Experimental-Result AVP
    // The Experimental-Result AVP indicates whether a particular
    // vendor specific request was completed successfully or whether
    // an error occurred.
    // The AVP Code of Experimental-Result AVP is 297.
    // The data type of Experimental-Result AVP is Grouped Type.
    ExperimentalResult experimentalResult;

    // Set the Vendor-Id AVP in Experimental-Result AVP.
    // The Vendor-Id AVP in the Experimental-Result AVP indicates
    // whether a particular vendor-specific request was completed
    // successfully or whether an error occurred.
    VendorId vendorId1(VENDOR_ID);
    experimentalResult.setVendorId(vendorId1);

    // Set the Experimental-Result-Code AVP.
    // The Experimental-Result-Code AVP contains a vendor-assigned
    // value representing the result of processing the request.
    // The AVP Code of AVP is 298 .
    // The data type of AVP is Unsigned32.
    ExperimentalResultCode experimentalResultCode(2001u);
    experimentalResult.setExperimentalResultCode(experimentalResultCode);

    // [2.2] Set the Experimental-Result AVP.

    // This AVP is required incase of Result-Code AVP is missing.
    CLAns.setExperimentalResult(experimentalResult);

    // [2.1.1] Set the Origin-Host AVP.
    // The Origin-Host AVP identifies the end-point that originated
    // the Diameter message.
    // The Origin-Host AVP must be present in all the Diameter
    // messages.
    // This AVP SHOULD be placed as close to the Diameter header as
    // possible.
    // The relay agents must not modify this AVP.
    // The AVP Code of Origin-Host AVP is 264
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginHost originHost(GetOriginHost());
    CLAns.setOriginHost(originHost);

    // [2.1.1] set the Origin-Realm AVP.
    // The Origin-Realm AVP contains the realm of the originator of
    // any Diameter message and must be present in all the Diameter
    // messages.
    // The AVP Code of Origin-Host AVP is 296.
    // The data type of Origin-Host AVP is DiameterIdentity.
    OriginRealm originRealm(GetOriginRealm());
    CLAns.setOriginRealm(originRealm);

    // The data type of Proxy-Info is Grouped Type.
    // This AVP should not be set by Application.
    // CLAns.setProxyInfo(diameter::base::ProxyInfo(...));

    // Create the Route-Record AVP.
    // The identity added in the Route-Record AVP must be the same as
    // the one received in the Origin-Host AVP of the message.
    // The AVP Code of Route-Record AVP is 282
    // The data type of Route-Record is DiameterIdentity.
    // This AVP should not be set by Application.
    // CLAns.setRouteRecord(diameter::base::RouteRecord(...));
    // add more AVP as ANY values

    if (appMode == UNIT_TEST)
    cout<<CLAns;

    // Send Object to stack.
    SendDiaAppMsg(CLAns, sIdx);

    // Update stats.
    S6AStats::GetS6AStats()->UpdateSendStats(CLAns.getCommandCode(), ITS_FALSE);
    return ITS_SUCCESS;
}

File: imsdia/api/java/test/diameter/s6a/CancelLocationTest.java
package diameter.s6a;

import com.intellinet.diameter.*;
import diameter.*;

public class CancelLocationTest {

    public static void testCancelLocationRequest() throws Exception {
        CancelLocationRequest cancelLocationRequest = new CancelLocationRequest(false, false);
        cancelLocationRequest.setSessionId(new diameter.base.SessionId("utf8string1067293343"));
        cancelLocationRequest.setAuthSessionState(new diameter.base.AuthSessionState(diameter.base.AuthSessionState.Value.NO_STATE_MAINTAINED));
        cancelLocationRequest.setOriginHost(new diameter.base.OriginHost("id-1088162968.diameter.com"));
        cancelLocationRequest.setOriginRealm(new diameter.base.OriginRealm("id-2028150358.diameter.com"));
        cancelLocationRequest.setDestinationHost(new diameter.base.DestinationHost("id-234988032.diameter.com"));
        cancelLocationRequest.setDestinationRealm(new diameter.base.DestinationRealm("id-1882818637.diameter.com"));
        cancelLocationRequest.setUserName(new diameter.base.UserName("utf8string2044292350"));
        cancelLocationRequest.setCancellationType(new CancellationType(CancellationType.Value.MME_UPDATE_PROCEDURE));
        {
            diameter.base.VendorSpecificApplicationId vendorSpecificApplicationId = new diameter.base.VendorSpecificApplicationId();
            vendorSpecificApplicationId.addVendorId(new diameter.base.VendorId(new Integer(1786006591)));
            vendorSpecificApplicationId.setAuthApplicationId(new diameter.base.AuthApplicationId(new Integer(541729315)));
            vendorSpecificApplicationId.setAcctApplicationId(new diameter.base.AcctApplicationId(new Integer(1206344418)));
            cancelLocationRequest.setVendorSpecificApplicationId(vendorSpecificApplicationId);
        }
        {
            SupportedFeatures supportedFeatures = new SupportedFeatures();
            supportedFeatures.setVendorId(new diameter.base.VendorId(new Integer(-639547713)));
            supportedFeatures.setFeatureListID(new FeatureListID(new Integer(-1832322847)));
            supportedFeatures.setFeatureList(new FeatureList(new Integer(177118476)));
            // add more AVP as ANY values to supportedFeatures
            cancelLocationRequest.addSupportedFeatures(supportedFeatures);
        }
        {
            diameter.base.ProxyInfo proxyInfo = new diameter.base.ProxyInfo();
            proxyInfo.setProxyHost(new diameter.base.ProxyHost("id-1806043815.diameter.com"));
            proxyInfo.setProxyState(new diameter.base.ProxyState("octetstring-1879873515"));
            // add more AVP as ANY values to proxyInfo
            cancelLocationRequest.addProxyInfo(proxyInfo);
        }
        cancelLocationRequest.addRouteRecord(new diameter.base.RouteRecord("id-1627806258.diameter.com"));
        // add more AVP as ANY values to cancelLocationRequest

        byte[] buffer = cancelLocationRequest.encode();
        CancelLocationRequest cancelLocationRequest2 = new CancelLocationRequest(buffer);
        String s1 = cancelLocationRequest.toString();
        String s2 = cancelLocationRequest2.toString();
        if (!s1.equals(s2)) {
            System.err.println("cancelLocationRequest1 = " + s1);
            System.err.println("cancelLocationRequest2 = " + s2);
            throw new VerificationException("CancelLocationRequest encoding/decoding failure");
        }
    }

    public static void testCancelLocationAnswer() throws Exception {
        CancelLocationAnswer cancelLocationAnswer = new CancelLocationAnswer(false, false);
        cancelLocationAnswer.setSessionId(new diameter.base.SessionId("utf8string222165565"));
        cancelLocationAnswer.setAuthSessionState(new diameter.base.AuthSessionState(diameter.base.AuthSessionState.Value.STATE_MAINTAINED));
        cancelLocationAnswer.setOriginHost(new diameter.base.OriginHost("id622261422.diameter.com"));
        cancelLocationAnswer.setOriginRealm(new diameter.base.OriginRealm("id-1174929849.diameter.com"));
        {
            diameter.base.VendorSpecificApplicationId vendorSpecificApplicationId = new diameter.base.VendorSpecificApplicationId();
            vendorSpecificApplicationId.addVendorId(new diameter.base.VendorId(new Integer(-1448703705)));
            vendorSpecificApplicationId.setAuthApplicationId(new diameter.base.AuthApplicationId(new Integer(-1919136974)));
            vendorSpecificApplicationId.setAcctApplicationId(new diameter.base.AcctApplicationId(new Integer(-991264878)));
            cancelLocationAnswer.setVendorSpecificApplicationId(vendorSpecificApplicationId);
        }
        {
            SupportedFeatures supportedFeatures = new SupportedFeatures();
            supportedFeatures.setVendorId(new diameter.base.VendorId(new Integer(-1849225961)));
            supportedFeatures.setFeatureListID(new FeatureListID(new Integer(-516791543)));
            supportedFeatures.setFeatureList(new FeatureList(new Integer(-499769161)));
            // add more AVP as ANY values to supportedFeatures
            cancelLocationAnswer.addSupportedFeatures(supportedFeatures);
        }
        cancelLocationAnswer.setResultCode(new diameter.base.ResultCode(new Integer(489440827)));
        {
            diameter.base.ExperimentalResult experimentalResult = new diameter.base.ExperimentalResult();
            experimentalResult.setVendorId(new diameter.base.VendorId(new Integer(-508239352)));
            experimentalResult.setExperimentalResultCode(new diameter.base.ExperimentalResultCode(new Integer(-108456628)));
            cancelLocationAnswer.setExperimentalResult(experimentalResult);
        }
        {
            diameter.base.ProxyInfo proxyInfo = new diameter.base.ProxyInfo();
            proxyInfo.setProxyHost(new diameter.base.ProxyHost("id258162430.diameter.com"));
            proxyInfo.setProxyState(new diameter.base.ProxyState("octetstring1437662110"));
            // add more AVP as ANY values to proxyInfo
            cancelLocationAnswer.addProxyInfo(proxyInfo);
        }
        cancelLocationAnswer.addRouteRecord(new diameter.base.RouteRecord("id1619303064.diameter.com"));
        // add more AVP as ANY values to cancelLocationAnswer

        byte[] buffer = cancelLocationAnswer.encode();
        CancelLocationAnswer cancelLocationAnswer2 = new CancelLocationAnswer(buffer);
        String s1 = cancelLocationAnswer.toString();
        String s2 = cancelLocationAnswer2.toString();
        if (!s1.equals(s2)) {
            System.err.println("cancelLocationAnswer1 = " + s1);
            System.err.println("cancelLocationAnswer2 = " + s2);
            throw new VerificationException("CancelLocationAnswer encoding/decoding failure");
        }
    }

    public static void main(String[] args) throws Exception {
        testCancelLocationRequest();
        testCancelLocationAnswer();
    }
}
